File: 2024/Python/13.py
data = open("../inputs/13.txt").read().splitlines()

claw_machines = []
newest_token = []
for num, line in enumerate(data):
    if num % 4 == 3:
        claw_machines.append(newest_token)
        newest_token = []
    else:
        newest_token.append(
            [*(int(x[2:]) for x in line.split(":")[1].strip().split(", "))]
        )
claw_machines.append(newest_token)

tokens_one = 0
tokens_two = 0

def solve_system(x1, y1, x2, y2, px, py):
    det = x1 * y2 - y1 * x2
    if det == 0:
        return None
    a = (px * y2 - py * x2) / det
    b = (x1 * py - y1 * px) / det
    if a.is_integer() and b.is_integer():
        return int(a), int(b)
    return None

for claw_machine in claw_machines:
    x1, y1 = claw_machine[0]
    x2, y2 = claw_machine[1]
    px, py = claw_machine[2]
    
    solution = solve_system(x1, y1, x2, y2, px, py)
    if solution:
        a, b = solution
        tokens_one += a * 3 + b

    offset = int(1e13)
    solution = solve_system(x1, y1, x2, y2, px + offset, py + offset)
    if solution:
        a, b = solution
        tokens_two += a * 3 + b

print(tokens_one)
print(tokens_two)

File: 2024/Deno/13.ts
const data = (await Deno.readTextFile("../inputs/13.txt")).split("\n");

const clawMachines: number[][][] = [];
let newestToken: number[][] = [];
for (let i = 0; i < data.length; i++) {
  if (i % 4 === 3) {
    clawMachines.push(newestToken);
    newestToken = [];
  } else {
    newestToken.push(
      data[i].split(":")[1].trim().split(", ").map((x) => parseInt(x.slice(2))),
    );
  }
}
clawMachines.push(newestToken);

let tokensOne = 0;
let tokensTwo = 0;

function solveSystem(
  x1: number,
  y1: number,
  x2: number,
  y2: number,
  px: number,
  py: number,
): [number, number] | null {
  const det = x1 * y2 - y1 * x2;
  if (det === 0) {
    return null;
  }
  const a = (px * y2 - py * x2) / det;
  const b = (x1 * py - y1 * px) / det;
  if (Number.isInteger(a) && Number.isInteger(b)) {
    return [Math.trunc(a), Math.trunc(b)];
  }
  return null;
}

for (const clawMachine of clawMachines) {
  if (clawMachine.length >= 3) {
    const [x1, y1] = clawMachine[0];
    const [x2, y2] = clawMachine[1];
    const [px, py] = clawMachine[2];

    const solution = solveSystem(x1, y1, x2, y2, px, py);
    if (solution) {
      const [a, b] = solution;
      tokensOne += a * 3 + b;
    }

    const offset = 1e13;
    const solution2 = solveSystem(x1, y1, x2, y2, px + offset, py + offset);
    if (solution2) {
      const [a, b] = solution2;
      tokensTwo += a * 3 + b;
    }
  }
}

console.log(tokensOne);
console.log(tokensTwo);
